google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/StreamWriter.java

```diff
@@ -599,13 +599,13 @@ private void doneCallback(Throwable finalStatus) {
               "Stream finished with error " + finalStatus.toString() + " for stream " + streamName);
         }
       }
+      Exceptions.StorageException storageException = Exceptions.toStorageException(finalStatus);
+      if (storageException != null) {
+        this.connectionFinalStatus = storageException;
+      }
     } finally {
       this.lock.unlock();
     }
-    Exceptions.StorageException storageException = Exceptions.toStorageException(finalStatus);
```

Comment:
why move this inside the lock?

Reply:
Actually, we should move it even up to l597, otherwise, there will be a problem.

Reply:
As discussed offline, the Exception conversion needs to be protected by the lock so that it is not thrown before getting converted.